### ISSUE TITLE
[RFC] Add RLVR-style Ranking Policy Loss to GRPO

### ISSUE THREAD

**Issue by li2zhi** [modelscope/ms-swift#8324]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new feature request. / 我已经搜索过现有的 issues，确认这是一个新的 Feature Request。

## 1. Motivation
Recent work such as [Rewards as Labels: Revisiting RLVR from a Classification Perspective](https://arxiv.org/pdf/2602.05630).

Compared with GRPO-style objectives currently supported in MS-Swift, RLVR introduces several advantages:
- does not require value function estimation
- avoids advantage estimation variance
- directly optimizes relative ranking between positive and negative responses
- better suited for binary / verifiable rewards

The objective can be summarized as:

<img width="630" height="136" alt="Image" src="https://github.com/user-attachments/assets/ad34f9dd-ecc6-427b-831c-276e6d01e19e" />

where responses are grouped by prompt.
This RFC proposes to introduce RLVR-style loss as a new optional loss type in GRPO.

### 2.1 New loss type
Add a new `loss_type`: 
```loss_type = "real"```

Implementation located under:
```swift/loss/real.py```

The loss operates at sequence level and requires grouping responses belonging to the same prompt.
Required inputs:
```
per_token_logps
old_per_token_logps
completion_mask
rewards
num_generations
```

### 2.2 Minimal Trainer Changes
The change to GRPOTrainer is minimal.

Add a branch in _compute_loss_and_metrics:
```
elif self.loss_type == "real":
    loss = real_loss_fn(...)
```

Add rewards into `batch_encoded_inputs`
```
        for batch, batch_encoded in zip(gas_chunks, batch_encoded_inputs):
            # Advantages are always [batch_size], will be broadcast to [batch_size, seq_len] in loss computation
            all_advantages = torch.stack([data['advantages'] for data in batch])
            batch_encoded['advantages'] = all_advantages

            all_rewards = torch.stack([data['rewards'] for data in batch])
            batch_encoded['rewards'] = all_rewards
```

## 3.Implementation Status
A prototype implementation has already been completed and integrated locally.Current features:
- RLVR ranking loss
- KL regularization support
- compatibility with GRPO rollout pipeline
- sequence-level score computation

However, one system-level issue arises during training.

RLVR requires **grouped responses per prompt**.Example:
```
prompt_i → num_generations responses

```

However, GRPO training uses:
```
per_device_train_batch_size
gradient_accumulation_steps
data parallel
```

The effective batch per loss computation becomes:
```local_batch = per_device_train_batch_size```
Example configuration:
```
per_device_train_batch_size = 2
gradient_accumulation_steps = 2
num_generations = 4
```

Rollout produces:
```
prompt A:
    response1
    response2
    response3
    response4
```
But `compute_loss` sees only:
```
step1:
    response1
    response2

step2:
    response3
    response4
```
Therefore, group is split across loss batches, which breaks the RLVR loss assumption.

At the moment, RLVR training works correctly only under the following constraint:
```
per_device_train_batch_size * dp_size must be divisible by num_generations

```
Do you have any good suggestions on how to minimize intrusion into the main workflow?

## Citation
[Rewards as Labels: Revisiting RLVR from a Classification Perspective](https://arxiv.org/pdf/2602.05630).
[REAL](https://github.com/DeepExperience/REAL).

**Comment by hjh0119** [comment]
> Terminology Clarification: RLVR (Reinforcement Learning with Verifiable Rewards) refers to a class of algorithms. The "RLVR" you mentioned should specifically refer to the **REAL** algorithm.

Overall looks good, a few suggestions:

1. No need to create a separate implementation under `swift/loss`. Instead:

- Implement as a method within `GRPOTrainer`, OR
- Implement directly in the `_compute_loss_and_metrics` method

2. Add `None` as an option for the `advantage_estimator` parameter. In this case, advantages equal the normalized rewards (Does this paper only support single-reward scenarios?)

**Comment by li2zhi** [comment]
Thanks for the suggestion.

My understanding of the paper is that it treats rewards as binary labels rather than scalar advantages. The objective partitions rollouts within each group into positive and negative sets based on the reward signal and then optimizes a ranking-style loss over the policy improvement score.

The reward primarily defines the label semantics (positive vs. negative samples). If we normalize rewards, may change the positive/negative partition within the group.

For multi-dimensional rewards, the current pipeline aggregates them through linear weighting before computing the final reward, which may weaken or discard part of the semantic information carried by each reward dimension.
